arangod/Aql/SharedQueryState.cpp

```diff
@@ -72,32 +72,33 @@ void SharedQueryState::waitForAsyncWakeup() {
     THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_KILLED);
   }
 
-  TRI_ASSERT(!_wakeupCb);
+  TRI_ASSERT(!_wakeupCallback);
   _cv.wait(guard, [&] { return _numWakeups > 0 || !_valid; });
   TRI_ASSERT(_numWakeups > 0 || !_valid);
-  _numWakeups--;
+  if (_valid) {
+    TRI_ASSERT(_numWakeups > 0);
+    --_numWakeups;
+  }
 }
 
 /// @brief setter for the continue handler:
 ///        We can either have a handler or a callback
 void SharedQueryState::setWakeupHandler(std::function<bool()> const& cb) {
   std::lock_guard<std::mutex> guard(_mutex);
-  _wakeupCb = cb;
+  // whenever we update the wakeupCallback, we also have to increase the
+  // callbackVersion and reset the numWakeups. Updating the callbackVersion is
+  // necessary to ensure that wakeup handlers that are still queued realize that
+  // they are no longer relevant (their associated RestHandler is already gone).
+  // Resetting the numWakeups is necessary to ensure that later wakeups actually
+  // schedule a new handler.
+  _wakeupCallback = cb;
   _numWakeups = 0;
-  _cbVersion++;
+  _callbackVersion++;
 }
 
 void SharedQueryState::resetWakeupHandler() {
   std::lock_guard<std::mutex> guard(_mutex);
-  _wakeupCb = nullptr;
-  _numWakeups = 0;
-  _cbVersion++;
-}
-
-void SharedQueryState::resetNumWakeups() {
-  std::lock_guard<std::mutex> guard(_mutex);
-  _numWakeups = 0;
-  _cbVersion++;
+  setWakeupHandler(nullptr);
```

Comment:
This calls `setWakeupHandler(nullptr);` while having acquired the mutex.
But `setWakeupHandler()` also acquires the mutex. I guess this will deadlock.

Reply:
Argl... yes, of course! 🤦‍♂️ That was one of the last cleanup changes I did.